File: packages/server/src/domains/groups/service.ts
import type { Group, CreateGroupInput, UpdateGroupInput, FlagGroup } from '@flick/shared';
import * as repo from './repo.js';
import * as flagEnvRepo from '../flag-environments/repo.js';
import { NotFoundError, ConflictError } from '../../shared/errors.js';
import { invalidateEnvFlags } from '../../shared/cache.js';

export async function listGroups(projectId: string): Promise<Group[]> {
  return repo.findByProject(projectId);
}

export async function getGroup(id: string): Promise<Group> {
  const group = await repo.findById(id);
  if (!group) throw new NotFoundError('Group', id);
  return group;
}

export async function createGroup(projectId: string, input: CreateGroupInput): Promise<Group> {
  const existing = await repo.findBySlug(projectId, input.slug);
  if (existing) throw new ConflictError(`Group with slug '${input.slug}' already exists`);
  return repo.create(projectId, input);
}

export async function updateGroup(id: string, input: UpdateGroupInput): Promise<Group> {
  const group = await repo.update(id, input);
  if (!group) throw new NotFoundError('Group', id);
  return group;
}

export async function deleteGroup(id: string): Promise<void> {
  const deleted = await repo.remove(id);
  if (!deleted) throw new NotFoundError('Group', id);
}

export async function getGroupsForFlagEnv(flagId: string, envId: string): Promise<Group[]> {
  const fe = await flagEnvRepo.findByFlagAndEnv(flagId, envId);
  if (!fe) throw new NotFoundError('FlagEnvironment');
  return repo.findGroupsForFlagEnv(fe.id);
}

export async function addGroupToFlag(flagId: string, envId: string, groupId: string): Promise<FlagGroup> {
  const fe = await flagEnvRepo.findByFlagAndEnv(flagId, envId);
  if (!fe) throw new NotFoundError('FlagEnvironment');

  const group = await repo.findById(groupId);
  if (!group) throw new NotFoundError('Group', groupId);

  const result = await repo.addGroupToFlag(fe.id, groupId);
  await invalidateEnvFlags(envId);
  return result;
}

export async function removeGroupFromFlag(flagId: string, envId: string, groupId: string): Promise<void> {
  const fe = await flagEnvRepo.findByFlagAndEnv(flagId, envId);
  if (!fe) throw new NotFoundError('FlagEnvironment');

  const removed = await repo.removeGroupFromFlag(fe.id, groupId);
  if (!removed) throw new NotFoundError('FlagGroup');
  await invalidateEnvFlags(envId);
}

File: packages/server-rust/src/domains/groups/service.rs
use fred::prelude::*;
use serde_json::Value;
use sqlx::PgPool;
use uuid::Uuid;

use crate::auth::AuthUser;
use crate::error::AppError;

use super::repo::{self, FlagGroupRow, GroupRow};

pub async fn list_groups(db: &PgPool, project_id: Uuid) -> Result<Vec<GroupRow>, AppError> {
    repo::find_by_project(db, project_id)
        .await
        .map_err(AppError::from)
}

pub async fn get_group(db: &PgPool, id: Uuid) -> Result<GroupRow, AppError> {
    repo::find_by_id(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("Group", Some(&id.to_string())))
}

pub async fn create_group(
    db: &PgPool,
    project_id: Uuid,
    name: &str,
    slug: &str,
    description: &str,
    rules: Value,
) -> Result<GroupRow, AppError> {
    // Check for slug uniqueness
    if let Some(_existing) = repo::find_by_slug(db, project_id, slug).await? {
        return Err(AppError::Conflict(format!(
            "Group with slug '{}' already exists",
            slug
        )));
    }

    repo::create(db, project_id, name, slug, description, rules)
        .await
        .map_err(AppError::from)
}

pub async fn update_group(
    db: &PgPool,
    id: Uuid,
    name: Option<&str>,
    description: Option<&str>,
    rules: Option<Value>,
) -> Result<GroupRow, AppError> {
    repo::update(db, id, name, description, rules)
        .await?
        .ok_or_else(|| AppError::not_found("Group", Some(&id.to_string())))
}

pub async fn delete_group(db: &PgPool, id: Uuid) -> Result<bool, AppError> {
    let deleted = repo::remove(db, id).await?;
    if !deleted {
        return Err(AppError::not_found("Group", Some(&id.to_string())));
    }
    Ok(deleted)
}

pub async fn get_groups_for_flag_env(
    db: &PgPool,
    flag_id: Uuid,
    env_id: Uuid,
) -> Result<Vec<GroupRow>, AppError> {
    let flag_env =
        crate::domains::flag_environments::repo::find_by_flag_and_env(db, flag_id, env_id)
            .await?
            .ok_or_else(|| AppError::not_found("FlagEnvironment", None))?;

    repo::find_groups_for_flag_env(db, flag_env.id)
        .await
        .map_err(AppError::from)
}

pub async fn add_group_to_flag(
    db: &PgPool,
    redis: &fred::clients::RedisPool,
    flag_id: Uuid,
    env_id: Uuid,
    group_id: Uuid,
) -> Result<FlagGroupRow, AppError> {
    // Verify flag environment exists
    let flag_env =
        crate::domains::flag_environments::repo::find_by_flag_and_env(db, flag_id, env_id)
            .await?
            .ok_or_else(|| AppError::not_found("FlagEnvironment", None))?;

    // Verify group exists
    repo::find_by_id(db, group_id)
        .await?
        .ok_or_else(|| AppError::not_found("Group", Some(&group_id.to_string())))?;

    let result = repo::add_group_to_flag(db, flag_env.id, group_id).await?;

    // Invalidate cache
    let cache_key = format!("flick:flag_env:{}:{}", flag_id, env_id);
    let _: Result<(), _> = redis.del::<(), _>(&cache_key).await;

    Ok(result)
}

pub async fn remove_group_from_flag(
    db: &PgPool,
    redis: &fred::clients::RedisPool,
    flag_id: Uuid,
    env_id: Uuid,
    group_id: Uuid,
) -> Result<bool, AppError> {
    // Verify flag environment exists
    let flag_env =
        crate::domains::flag_environments::repo::find_by_flag_and_env(db, flag_id, env_id)
            .await?
            .ok_or_else(|| AppError::not_found("FlagEnvironment", None))?;

    let removed = repo::remove_group_from_flag(db, flag_env.id, group_id).await?;
    if !removed {
        return Err(AppError::not_found("FlagGroup", None));
    }

    // Invalidate cache
    let cache_key = format!("flick:flag_env:{}:{}", flag_id, env_id);
    let _: Result<(), _> = redis.del::<(), _>(&cache_key).await;

    Ok(removed)
}
